// @softtime/mobile/src/entities/request/api/mock.ts
import { RequestType, RequestStatus } from '@softtime/shared';
import type { AbsenceRequest } from '@softtime/shared';

export const mockMyRequests: AbsenceRequest[] = [
  {
    id: 'req-001',
    companyId: 'company-001',
    userId: 'user-worker-001',
    type: RequestType.SICK,
    startDate: new Date('2026-06-02'),
    endDate: new Date('2026-06-03'),
    desiredTime: null,
    comment: 'Болею, есть справка',
    status: RequestStatus.APPROVED,
    decidedBy: 'user-admin-001',
    decisionNote: null,
    createdAt: new Date('2026-06-01T18:30:00'),
  },
  {
    id: 'req-002',
    companyId: 'company-001',
    userId: 'user-worker-001',
    type: RequestType.VACATION,
    startDate: new Date('2026-07-01'),
    endDate: new Date('2026-07-14'),
    desiredTime: null,
    comment: 'Ежегодный отпуск',
    status: RequestStatus.PENDING,
    decidedBy: null,
    decisionNote: null,
    createdAt: new Date('2026-06-06T10:00:00'),
  },
  {
    id: 'req-003',
    companyId: 'company-001',
    userId: 'user-worker-001',
    type: RequestType.EARLY_LEAVE,
    startDate: new Date('2026-05-28'),
    endDate: null,
    desiredTime: '16:00',
    comment: 'Врач в 16:30',
    status: RequestStatus.REJECTED,
    decidedBy: 'user-admin-001',
    decisionNote: 'Нет свободных специалистов для покрытия',
    createdAt: new Date('2026-05-27T09:00:00'),
  },
];

export const mockIncomingRequests: AbsenceRequest[] = [
  {
    id: 'req-incoming-001',
    companyId: 'company-001',
    userId: 'user-worker-002',
    type: RequestType.REMOTE,
    startDate: new Date('2026-06-10'),
    endDate: new Date('2026-06-10'),
    desiredTime: null,
    comment: 'Работаю удалённо по семейным обстоятельствам',
    status: RequestStatus.PENDING,
    decidedBy: null,
    decisionNote: null,
    createdAt: new Date('2026-06-06T12:00:00'),
  },
  {
    id: 'req-incoming-002',
    companyId: 'company-001',
    userId: 'user-worker-001',
    type: RequestType.VACATION,
    startDate: new Date('2026-07-01'),
    endDate: new Date('2026-07-14'),
    desiredTime: null,
    comment: 'Ежегодный отпуск',
    status: RequestStatus.PENDING,
    decidedBy: null,
    decisionNote: null,
    createdAt: new Date('2026-06-06T10:00:00'),
  },
  {
    id: 'req-incoming-003',
    companyId: 'company-001',
    userId: 'user-worker-002',
    type: RequestType.SICK,
    startDate: new Date('2026-06-08'),
    endDate: new Date('2026-06-09'),
    desiredTime: null,
    comment: 'Болею, есть справка от врача',
    status: RequestStatus.APPROVED,
    decidedBy: 'user-admin-001',
    decisionNote: null,
    createdAt: new Date('2026-06-05T14:30:00'),
  },
];
